### ISSUE TITLE
Bad RTSP status code with GStreamer

### ISSUE THREAD

**Issue by zoolyka** [mbullington/yellowstone#33]
Great project! I'm trying to make the wowza example work with a GStreamer RTSP server, but always receiving the following error:

`(node:7024) UnhandledPromiseRejectionWarning: Error: Bad RTSP status code 460!`

_460 - Only aggregate operation allowed_

Any guess what could be wrong on either side? Thank you.

**Comment by zoolyka** [comment]
Sure, the full log:

> zooly@dev18:~/yellowstone/examples$ node wowza.js
> C->S: OPTIONS rtsp://192.168.0.248:8554/stream RTSP/1.0
> CSeq: 1
> User-Agent: yellowstone/3.x
> 
> S->C: RTSP/1.0 200 OK
> CSeq: 1
> Public: OPTIONS, DESCRIBE, ANNOUNCE, GET_PARAMETER, PAUSE, PLAY, RECORD, SETUP, SET_PARAMETER, TEARDOWN
> Server: GStreamer RTSP server
> Date: Wed, 28 Aug 2019 17:59:47 GMT
> 
> 
> C->S: DESCRIBE rtsp://192.168.0.248:8554/stream RTSP/1.0
> CSeq: 2
> User-Agent: yellowstone/3.x
> Accept: application/sdp
> 
> S->C: v=0
> o=- 2264366194808406837 1 IN IP4 192.168.0.248
> s=Session streamed with GStreamer
> i=rtsp-server
> t=0 0
> a=tool:GStreamer
> a=type:broadcast
> a=control:*
> a=range:npt=now-
> m=video 0 RTP/AVP 96
> c=IN IP4 0.0.0.0
> a=rtpmap:96 H264/90000
> a=framerate:60
> a=fmtp:96 packetization-mode=1;profile-level-id=424015;sprop-parameter-sets=Z0JAKJWgFAFuQA==,aM48gA==
> a=control:stream=0
> a=ts-refclk:local
> a=mediaclk:sender
> v=0
> o=- 2264366194808406837 1 IN IP4 192.168.0.248
> s=Session streamed with GStreamer
> i=rtsp-server
> t=0 0
> a=tool:GStreamer
> a=type:broadcast
> a=control:*
> a=range:npt=now-
> m=video 0 RTP/AVP 96
> c=IN IP4 0.0.0.0
> a=rtpmap:96 H264/90000
> a=framerate:60
> a=fmtp:96 packetization-mode=1;profile-level-id=424015;sprop-parameter-sets=Z0JAKJWgFAFuQA==,aM48gA==
> a=control:stream=0
> a=ts-refclk:local
> a=mediaclk:sender
> 
> C->S: SETUP rtsp://192.168.0.248:8554/stream/stream=0 RTSP/1.0
> CSeq: 3
> User-Agent: yellowstone/3.x
> Transport: RTP/AVP/TCP;interleaved=0-1
> 
> S->C: RTSP/1.0 200 OK
> CSeq: 3
> Transport: RTP/AVP/TCP;unicast;interleaved=0-1;ssrc=CF588B1E;mode="PLAY"
> Server: GStreamer RTSP server
> Session: Bwai+coDnS-txpdu
> Date: Wed, 28 Aug 2019 17:59:48 GMT
> 
> 
> Connected. Video format is H264
> C->S: PLAY rtsp://192.168.0.248:8554/stream/stream=0 RTSP/1.0
> CSeq: 4
> User-Agent: yellowstone/3.x
> Session: Bwai+coDnS-txpdu
> 
> S->C: RTSP/1.0 460 Only aggregate operation allowed
> CSeq: 4
> Server: GStreamer RTSP server
> Date: Wed, 28 Aug 2019 17:59:48 GMT
> 
> 
> (node:7119) UnhandledPromiseRejectionWarning: Error: Bad RTSP status code 460!
>     at RTSPClient.responseHandler (/home/zooly/yellowstone/dist/RTSPClient.js:256:28)
>     at emitThree (events.js:141:20)
>     at RTSPClient.emit (events.js:217:7)
>     at RTSPClient._onData (/home/zooly/yellowstone/dist/RTSPClient.js:403:30)
>     at emitOne (events.js:116:13)
>     at Socket.emit (events.js:211:7)
>     at addChunk (_stream_readable.js:263:12)
>     at readableAddChunk (_stream_readable.js:250:11)
>     at Socket.Readable.push (_stream_readable.js:208:10)
>     at TCP.onread (net.js:607:20)
> (node:7119) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
> (node:7119) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ^C

Thank you.

**Comment by zoolyka** [comment]
Just found this related topic: https://stackoverflow.com/questions/8699027/extract-audio-from-rtsp-link-for-android

So maybe (in the above case) it should send SETUP and PLAY for simply "/stream" insted of "/stream/stream=0" - if thats the way to request an aggregated stream?

In my understanding appending that control variable to the URL means that you want a specific part (audio/video/etc.) part of a stream - which is a non-aggregated call.

**Comment by zoolyka** [comment]
Was able to solve it.

![image](https://user-images.githubusercontent.com/4547445/63940610-28b7f780-ca6a-11e9-8861-a20b123bb82c.png)

**Comment by mbullington** [comment]
Great catch! I'll take a closer look and add this to the repo soon.

**Comment by RogerHardiman** [comment]
The root cause of the problem is now fixed in commit 2028c2aabc2726c0927be4fde84edee04677c3c5 so you should not need the new `aggregate` flag. However your workaround correctly identified that the problem was Yellowstone mixed up Aggregate URLs and Media Stream URLS.

EG
```
Aggregate URL = rtsp://1.2.3.4/cats
Video Stream Media URL = rtsp://1.2.3.4/cats/streamID=0 
Audio Stream Media URL = rtsp://1.2.3.4/cats/streamID=1 
Meta data Media URL = rtsp://1.2.3.4/cats/streamID=1 
```

Normally we use Aggregate in DESCRIBE, PLAY, TEARDOWN and the media stream URLs in SETUP
